huxley/invoice_automation/src/util/quick_books_utils.py

```diff
@@ -0,0 +1,276 @@
+    :return: Converted Address object, None if None is passed
+    """
+    if address is None:
+        return None
+
+    return Address(
+        address.Line1,
+        address.Line2,
+        address.City,
+        address.CountrySubDivisionCode,
+        address.Country,
+        address.PostalCode
+    )
+
+
+def check_invoice_matches_items_and_counts(invoice: Invoice, item_names: List[str], item_counts: List[int]) -> bool:
+    """
+    Checks that the passed invoice match the passed items and counts
+    Invoice should consist exactly of items with correct counts
+    item_names[i] is expected to have count item_counts[i]
+    Only SalesItemLines are checked from the invoice
+    :param invoice:
+    :param item_names:
+    :param item_counts:
+    :return:
+    """
+    if len(item_names) != len(item_counts):
+        raise ValueError("item_refs and item_counts were expected to have the same length")
+
+    line_items = list(filter(lambda i: i.DetailType == SALES_ITEM_LINE_DETAIL_TYPE, invoice.Line))
+    if len(line_items) != len(item_names):
+        return False
+
+    invoice_item_counts = set()
+    for item in line_items:
+        invoice_item_name = item.SalesItemLineDetail.ItemRef.name
+        # the item name in the invoice could have a category prepended
+        if ":" in invoice_item_name:
+            invoice_item_name = invoice_item_name.split(":")[1]
+        invoice_item_count = item.SalesItemLineDetail.Qty
+        invoice_item_counts.add((invoice_item_name, invoice_item_count))
+
+    ref_item_counts = {item_count_tuple for item_count_tuple in zip(item_names, item_counts)}
+
+    return invoice_item_counts == ref_item_counts
+
+
+SALES_ITEM_LINE_DETAIL_TYPE = "SalesItemLineDetail"
+
+
+SERVICE_DATES = {
+    Conference.BMUN71: datetime.date(2023, 3, 5),
+    Conference.FC: datetime.date(2022, 11, 12)
+}
+
+
+def create_SalesItemLine(item: Item, quantity: int, service_date: datetime.date) -> SalesItemLine:
+    """
+    Creates a SalesItemLine for use in an invoice
+    :param item:
+    :param quantity:
+    :param service_date:
+    :return:
+    """
+    line = SalesItemLine()
+    line.Amount = int(quantity) * item.UnitPrice
+    detail = SalesItemLineDetail()
+    detail.ItemRef = item.to_ref()
+    detail.Qty = quantity
+    detail.UnitPrice = item.UnitPrice
+    detail.ServiceDate = service_date.isoformat()
+    line.SalesItemLineDetail = detail
+    return line
+
+
+def create_credit_card_fee_SalesItemLine(amount: float, cc_fee_item: Item) -> SalesItemLine:
+    """
+
+    :param cc_fee_item:
+    :param amount: Pre-tax amount
+    :return:
+    """
+    fee_amount = get_adjustment_fee_from_subtotal(amount)
+    line = SalesItemLine()
+    line.Amount = fee_amount
+
+    detail = SalesItemLineDetail()
+    detail.ItemRef = cc_fee_item.to_ref()
+    detail.ServiceDate = datetime.date.today().isoformat()
+    line.SalesItemLineDetail = detail
+
+    return line
+
+
+END_DATES = {
```

Comment:
This file seems to have a lot of hardcoding of dates and is specific to BMUN 71. Can we generalize this again based on the conference model?

Reply:
Hardcoding based on the conference model means that these won't be available when trying to run scripts with the invoice automation library offline. Can we make this change in a future change once all registration invoices are fully automated?

Reply:
Resolving: Changes will be made in separate PR